### PR TITLE
Ncar isca

### PR DESCRIPTION
**Pull request by wanyingkang**
Dear Steven,

Sorry for the delay. I took a 1-week trip in Italy before came back. 
This is the ncar machine version. Please let me know if I didn't do it correctly.

Best,
Wanying

### CODE REVIEW

**Comment by jamesp** [comment]
Hi Wanying,

Thanks for your pull request! It's great that you got Isca running at NCAR, and we really like the documentation you put together.

To merge this in, we'll need to make a few changes, mainly putting as much of the NCAR specific bits into the environment file.  I'll put comments alongside the changes.

Thanks again,
James

File: src/extra/loadmodule [hunk 1]
```diff
@@ -1,6 +1,10 @@
 echo loadmodules
 
 module purge
-module load hdf5/1.8.11  
-module load netcdf-4.3.3.1-openmpi-intel
-module load git/2.1.2
+module load ncarenv/1.2
```
Comment (jamesp) [inline src/extra/loadmodule]: Could you please revert these changes to `src/extra/loadmodule` as they are redundant when using the `src/extra/env/cheyenne` environment.

File: src/extra/python/isca/codebase.py [hunk 2]
```diff
@@ -30,7 +30,7 @@ def from_repo(cls, repo, commit=None, **kwargs):
         return cls(repo=repo, commit=commit, **kwargs)
 
     @classmethod
-    def from_directory(cls, directory, **kwargs):
+    def from_directory(cls, directory,**kwargs):
```
Comment (jamesp) [inline src/extra/python/isca/codebase.py]: Please add the space back in here

File: src/extra/python/isca/codebase.py [hunk 3]
```diff
@@ -217,7 +217,9 @@ def _log_line(self, line):
 
     @useworkdir
     @destructive
-    def compile(self, debug=False, optimisation=None):
+    # def compile(self, debug=False, optimisation=None):
```
Comment (jamesp) [inline src/extra/python/isca/codebase.py]: Please could you remove the changes for supporting source modifications for now?  I think it is a good idea, but we must be careful how we implement it so as not to break other people's existing experiment scripts.  We should open a separate issue to track and implement this.

**Comment by wanyingkang** [comment]
Hi James,

Thank you for taking a look at my pull request. Recently NCAR modified some basic settings and paths in their machine, and now even myself cannot run isca on cheyenne. I will try to work on that ASAP and get back to you.
As for making changes, please feel free to go ahead and do it. 

Many thanks!

Best,
Wanying

File: src/extra/python/isca/experiment.py [hunk 4]
```diff
@@ -307,6 +307,13 @@ def _outhandler(line):
                 self.log.debug("Restart file %s combined" % restartfile)
 
             self.emit('run:combined', self)
+        else:
+            for file in self.diag_table.files:
```
Comment (jamesp) [inline src/extra/python/isca/experiment.py]: This should be moved to `isca.util.py` as e.g. 
```{python}
def move_output(exp):
     for file in exp.diag_table.files:
         ....
```

File: src/extra/python/isca/templates/compile.sh [hunk 5]
```diff
@@ -1,16 +1,15 @@
 #!/usr/bin/env bash
 # Compiles the MiMa GFDL Moist Model (tested on emps-gv2.ex.ac.uk)
 
-
 # 1. Configuration
 hostname=`hostname`
-template={{ template_dir }}/mkmf.template.ia64
+template={{ template_dir }}/mkmf.cheyenne.intel
```
Comment (jamesp) [inline src/extra/python/isca/templates/compile.sh]: This `compile.sh` file is used for all the different architectures so we must not hard code cheyenne or any other specific info in here.

File: src/extra/python/isca/templates/compile.sh [hunk 6]
```diff
@@ -20,6 +19,7 @@ netcdf_flags=`nc-config --fflags --flibs`
 # 2. Load the necessary tools into the environment
 module purge
 source {{ env_source }}
+source $python_package_lib_path
```
Comment (jamesp) [inline src/extra/python/isca/templates/compile.sh]: What is contained within this source that is not possible to put into the environment file `src/extra/env/cheyenne`?

File: src/extra/python/isca/templates/mkmf.cheyenne.intel [hunk 7]
```diff
@@ -0,0 +1,9 @@
+# template for the Intel fortran compiler version 8
```
Comment (jamesp) [inline src/extra/python/isca/templates/mkmf.cheyenne.intel]: How different is this file from the default mkmf template?  Could it be achieved by putting things into the environment file instead?

**Comment by jamesp** [comment]
OK, let me know when you have a set of config that does allow isca to run on NCAR again and I'll help you to put it into a PR, and then into the main repo. cheers!

**Comment by wanyingkang** [comment]
Hi James,

I managed to run isca again on Cheyenne. Instructions are written in README.cheyenne. Now, let’s discuss what I can do to help merge my version to main branch.
I am new to github, so if you could briefly describe how I can fetch your comments and respond to them, that will be very helpful.
I got your general point is to maintain everything in the same framework, which makes sense and will make other people’s life easier. I will try to adjust the code toward that direction.

Best,
Wanying

**Comment by wanyingkang** [comment]
Hi James,

I proposed the following changes. 

> On Jul 30, 2018, at 10:41 AM, James Penn <notifications@github.com> wrote:
> 
> @jamesp commented on this pull request.
> 
> In src/extra/loadmodule <https://github.com/ExeClim/Isca/pull/55#discussion_r206156069>:
> 
> > @@ -1,6 +1,10 @@
>  echo loadmodules
>  
>  module purge
> -module load hdf5/1.8.11  
> -module load netcdf-4.3.3.1-openmpi-intel
> -module load git/2.1.2
> +module load ncarenv/1.2
> Could you please revert these changes to src/extra/loadmodule as they are redundant when using the src/extra/env/cheyenne environment.
> 
May I replace all these with source ${GFDL_BASE}/src/extra/env/${GFDL_ENV}? 
Actually this is not necessary. Just for my convenience, so you can ignore this file, or I can commit a checkpoint without this file.

> 
> In src/extra/python/isca/codebase.py <https://github.com/ExeClim/Isca/pull/55#discussion_r206156238>:
> 
> > @@ -30,7 +30,7 @@ def from_repo(cls, repo, commit=None, **kwargs):
>          return cls(repo=repo, commit=commit, **kwargs)
>  
>      @classmethod
> -    def from_directory(cls, directory, **kwargs):
> +    def from_directory(cls, directory,**kwargs):
> Please add the space back in here
> 
> In src/extra/python/isca/codebase.py <https://github.com/ExeClim/Isca/pull/55#discussion_r206156917>:
> 
> > @@ -217,7 +217,9 @@ def _log_line(self, line):
>  
>      @useworkdir
>      @destructive
> -    def compile(self, debug=False, optimisation=None):
> +    # def compile(self, debug=False, optimisation=None):
> Please could you remove the changes for supporting source modifications for now? I think it is a good idea, but we must be careful how we implement it so as not to break other people's existing experiment scripts. We should open a separate issue to track and implement this.
> 
I can commit a checkpiont without this file

> 
> In src/extra/python/isca/experiment.py <https://github.com/ExeClim/Isca/pull/55#discussion_r206158038>:
> 
> > @@ -307,6 +307,13 @@ def _outhandler(line):
>                  self.log.debug("Restart file %s combined" % restartfile)
>  
>              self.emit('run:combined', self)
> +        else:
> +            for file in self.diag_table.files:
> This should be moved to isca.util.py as e.g.
> 
> def move_output(exp):
>      for file in exp.diag_table.files:
>          ….
I do not recognize this change, and it is not marked with wykang so I do not think I made these changes. I will exclude this file in the new commit.
> 
> In src/extra/python/isca/templates/compile.sh <https://github.com/ExeClim/Isca/pull/55#discussion_r206159085>:
> 
> >  # 1. Configuration
>  hostname=`hostname`
> -template={{ template_dir }}/mkmf.template.ia64
> +template={{ template_dir }}/mkmf.cheyenne.intel
> This compile.sh file is used for all the different architectures so we must not hard code cheyenne or any other specific info in here.
> 
I tested the mkmf.template.ia64. It also works.
> 
> In src/extra/python/isca/templates/compile.sh <https://github.com/ExeClim/Isca/pull/55#discussion_r206159526>:
> 
> > @@ -20,6 +19,7 @@ netcdf_flags=`nc-config --fflags --flibs`
>  # 2. Load the necessary tools into the environment
>  module purge
>  source {{ env_source }}
> +source $python_package_lib_path
> What is contained within this source that is not possible to put into the environment file src/extra/env/cheyenne?
> 
I merged it into src/extra/env/cheyenne. This is better.
> 
> In src/extra/python/isca/templates/mkmf.cheyenne.intel <https://github.com/ExeClim/Isca/pull/55#discussion_r206161035>:
> 
> > @@ -0,0 +1,9 @@
> +# template for the Intel fortran compiler version 8
> How different is this file from the default mkmf template? Could it be achieved by putting things into the environment file instead?
> 
They are similar. I just used the template I use for porting MiMA as I was sure it would work.  Now I excluded the cheyenne template.

**Comment by wanyingkang** [comment]
> On Jul 30, 2018, at 5:47 PM, Wanying Kang <kangwanying1992@gmail.com> wrote:
> 
> Hi James,
> 
> I proposed the following changes. 
> 
>> On Jul 30, 2018, at 10:41 AM, James Penn <notifications@github.com <mailto:notifications@github.com>> wrote:
>> 
>> @jamesp commented on this pull request.
>> 
>> In src/extra/loadmodule <https://github.com/ExeClim/Isca/pull/55#discussion_r206156069>:
>> 
>> > @@ -1,6 +1,10 @@
>>  echo loadmodules
>>  
>>  module purge
>> -module load hdf5/1.8.11  
>> -module load netcdf-4.3.3.1-openmpi-intel
>> -module load git/2.1.2
>> +module load ncarenv/1.2
>> Could you please revert these changes to src/extra/loadmodule as they are redundant when using the src/extra/env/cheyenne environment.
>> 
> May I replace all these with source ${GFDL_BASE}/src/extra/env/${GFDL_ENV}? 
> Actually this is not necessary. Just for my convenience, so you can ignore this file, or I can commit a checkpoint without this file.
> 
>> 
>> In src/extra/python/isca/codebase.py <https://github.com/ExeClim/Isca/pull/55#discussion_r206156238>:
>> 
>> > @@ -30,7 +30,7 @@ def from_repo(cls, repo, commit=None, **kwargs):
>>          return cls(repo=repo, commit=commit, **kwargs)
>>  
>>      @classmethod
>> -    def from_directory(cls, directory, **kwargs):
>> +    def from_directory(cls, directory,**kwargs):
>> Please add the space back in here
>> 
>> In src/extra/python/isca/codebase.py <https://github.com/ExeClim/Isca/pull/55#discussion_r206156917>:
>> 
>> > @@ -217,7 +217,9 @@ def _log_line(self, line):
>>  
>>      @useworkdir
>>      @destructive
>> -    def compile(self, debug=False, optimisation=None):
>> +    # def compile(self, debug=False, optimisation=None):
>> Please could you remove the changes for supporting source modifications for now? I think it is a good idea, but we must be careful how we implement it so as not to break other people's existing experiment scripts. We should open a separate issue to track and implement this.
>> 
> I can commit a checkpiont without this file
> 
>> 
>> In src/extra/python/isca/experiment.py <https://github.com/ExeClim/Isca/pull/55#discussion_r206158038>:
>> 
>> > @@ -307,6 +307,13 @@ def _outhandler(line):
>>                  self.log.debug("Restart file %s combined" % restartfile)
>>  
>>              self.emit('run:combined', self)
>> +        else:
>> +            for file in self.diag_table.files:
>> This should be moved to isca.util.py as e.g.
>> 
>> def move_output(exp):
>>      for file in exp.diag_table.files:
>>          ….
> I do not recognize this change, and it is not marked with wykang so I do not think I made these changes. I will exclude this file in the new commit.
Now I recall this modification. I did this because of a bug. The output files will be combined and moved only if there are multiple processors, and if one uses 1 processor, the file will be overwritten by next run without saving it to the data directory. But I could not find isca.util.py in the folder… Did you just added it? If so, maybe you can merge?

**Comment by jamesp** [comment]
> May I replace all these with source ${GFDL_BASE}/src/extra/env/${GFDL_ENV}?

That seems like a good idea, yes!

> Now I recall this modification. I did this because of a bug. The output files will be combined and moved only if there are multiple processors, and if one uses 1 processor, the file will be overwritten by next run without saving it to the data directory. But I could not find isca.util.py in the folder… Did you just added it? If so, maybe you can merge?

You're right, sorry I now see this is also in other branches of the codebase and it is needed.

File: src/extra/python/isca/templates/compile.sh (hunk 6 above)
Comment (jamesp) [reply src/extra/python/isca/templates/compile.sh]: The issue is this line change in `compile.sh`.  Could you remove it, and do the $python_package_lib_path thing in the NCAR specific env file?  It should have the same effect as this file is sourced in the line above

**Comment by wanyingkang** [comment]
Yes, I did that.

Best regards,
Wanying

**Comment by jamesp** [comment]
I think this is looking pretty good now.  

@wanyingkang the `build_isca.py` script currently seems very specific to you and the experiments you are doing.  There are several stages of processing you are doing, e.g. lines 12 to 50 that seem to be recreating a lot of what the python framework already gives you over making directories etc.

You might be interested in the `run_cli` method that is going to go into the master codebase with #45.  It will allow you to do a lot of what you are doing in your file `build_isca.py` to allow for command line arguments.  For example, any of the test_cases could be changed like this:

2. Don't always perform the compilation (normally around line 22 of test cases).  Comment out this line:
```python
# cb.compile()  # compile the source code to working directory $GFDL_WORK/codebase
```
1. Remove this section:
```python
#Lets do a run!
if __name__=="__main__":
    exp.run(1, use_restart=False, num_cores=NCORES)
    for i in range(2,121):
        exp.run(i, num_cores=NCORES)
```

2. Replace with this:
```python
from isca.util import run_cli
run_cli(exp)
```

3. From the command line, you can now control your experiments
To compile:
`python exp_file.py --compile`
To run:
`python exp_file.py --up-to --run 20 --num-cores 16`

**Comment by jamesp** [comment]
Brilliant, thank you for your input, we really appreciate it.   I'm sure it will help many others at NCAR get up and running quickly, which is great!
